src/main/java/me/jellysquid/mods/sodium/mixin/features/buffer_builder/fast_advance/MixinBufferBuilder.java

```diff
@@ -20,24 +26,36 @@ public abstract class MixinBufferBuilder extends FixedColorVertexConsumer implem
     @Shadow
     private int currentElementId;
 
+    @Unique
+    private VertexFormatElement[] formatElements;
+
+    @Inject(method = "setFormat",
+        at = @At(
+            value = "FIELD",
+            target = "Lnet/minecraft/client/render/BufferBuilder;format:Lnet/minecraft/client/render/VertexFormat;",
+            opcode = Opcodes.PUTFIELD
+        )
+    )
+    private void onFormatChanged(VertexFormat format, CallbackInfo ci) {
+        this.formatElements = format.getElements().toArray(new VertexFormatElement[0]);
```

Comment:
I don't particularly like the memory allocation here, but I'm not sure whether this function is called enough to matter...

Maybe we can store a `VertexFormatElement[32]` in the `BufferBuilder`, and then re-use the array? OpenGL only allows a maximum of 32 vertex attributes. We'd just need to store the "actual" number of vertex elements alongside the array.

Furthermore, we could strip the `PADDING` elements so that `BufferBuilder.nextElement` doesn't need to check & skip them. That might result in better code generation.